plugins/org.eclipse.emf.ecore/src/org/eclipse/emf/ecore/plugin/EcorePlugin.java

```diff
@@ -1309,4 +1337,361 @@ private static void computeModels(Map<URI, URI> pluginMap, Map<String, URI> nsUR
+      {
+        readLock.unlock();
+      }
+    }
+
+    private <U, R> R whileLocked(BiFunction<EPackage.Registry, U, R> operation, U value)
+    {
+
+      ReadLock readLock = lock.readLock();
+      readLock.lock();
+      try
+      {
+        return operation.apply(getEffectiveRegistry(), value);
+      }
+      finally
+      {
+        readLock.unlock();
+      }
+    }
+
+    @Override
+    public int size()
+    {
+      return whileLocked(EPackage.Registry::size);
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+      return whileLocked(EPackage.Registry::isEmpty);
+    }
+
+    @Override
+    public boolean containsKey(Object key)
+    {
+      return whileLocked(EPackage.Registry::containsKey, key);
+    }
+
+    @Override
+    public boolean containsValue(Object value)
+    {
+      return whileLocked(EPackage.Registry::containsValue, value);
+    }
+
+    @Override
+    public Object get(Object key)
+    {
+      return whileLocked(EPackage.Registry::get, key);
+    }
+
+    @Override
+    public Object put(String key, Object value)
+    {
+      ReadLock readLock = lock.readLock();
+      readLock.lock();
+      try
+      {
+        if (delegate != null)
+        {
+          delegate.put(key, value);
+        }
+        return backup.put(key, value);
+      }
+      finally
+      {
+        readLock.unlock();
+      }
+    }
+
+    @Override
+    public Object remove(Object key)
+    {
+      ReadLock readLock = lock.readLock();
+      readLock.lock();
+      try
+      {
+        if (delegate != null)
+        {
+          delegate.remove(key);
+        }
+        return backup.remove(key);
+      }
+      finally
+      {
+        readLock.unlock();
+      }
+    }
+
+    @Override
+    public void putAll(Map<? extends String, ? extends Object> m)
+    {
+      ReadLock readLock = lock.readLock();
+      readLock.lock();
+      try
+      {
+        if (delegate != null)
+        {
+          delegate.putAll(m);
+        }
+        backup.putAll(m);
+      }
+      finally
+      {
+        readLock.unlock();
+      }
+    }
+
+    @Override
+    public void clear()
+    {
+      ReadLock readLock = lock.readLock();
+      readLock.lock();
+      try
+      {
+        if (delegate != null)
+        {
+          delegate.clear();
+        }
+        backup.clear();
+      }
+      finally
+      {
+        readLock.unlock();
+      }
+    }
+
+    @Override
+    public Set<String> keySet()
+    {
+      return whileLocked(() -> Collections.unmodifiableSet(new LinkedHashSet<>(getEffectiveRegistry().keySet())));
+    }
+
+    @Override
+    public Collection<Object> values()
+    {
+      return whileLocked(() -> Collections.unmodifiableCollection(new ArrayList<>(getEffectiveRegistry().values())));
+    }
+
+    @Override
+    public Set<Entry<String, Object>> entrySet()
+    {
+      return whileLocked(() -> Collections.unmodifiableSet(new LinkedHashMap<>(getEffectiveRegistry()).entrySet()));
+    }
```

Comment:
Note that I made these views effectively be read-only snapshot copies.

I expect nothing is relying on these being views that track underlying changes nor relying on modifying the registry via these views.